state-sync/storage-service/server/src/lib.rs

```diff
@@ -65,6 +72,134 @@ impl Error {
     }
 }
 
+/// A subscription for data received by a client
+pub struct DataSubscriptionRequest {
+    protocol: ProtocolId,
+    request: StorageServiceRequest,
+    response_sender: ResponseSender,
+    subscription_start_time: Instant,
+    time_service: TimeService,
+}
+
+impl DataSubscriptionRequest {
+    fn new(
+        protocol: ProtocolId,
+        request: StorageServiceRequest,
+        response_sender: ResponseSender,
+        time_service: TimeService,
+    ) -> Self {
+        Self {
+            protocol,
+            request,
+            response_sender,
+            subscription_start_time: time_service.now(),
+            time_service,
+        }
+    }
+
+    /// Creates a new storage service request to satisfy the transaction
+    /// subscription using the new data at the specified `target_ledger_info`.
+    fn get_storage_request_for_missing_data(
+        &self,
+        config: StorageServiceConfig,
+        target_ledger_info: &LedgerInfoWithSignatures,
+    ) -> Result<StorageServiceRequest, Error> {
+        // Calculate the number of versions to fetch
+        let known_version = self.highest_known_version();
+        let target_version = target_ledger_info.ledger_info().version();
+        let mut num_versions_to_fetch =
+            target_version.checked_sub(known_version).ok_or_else(|| {
+                Error::UnexpectedErrorEncountered(
+                    "Number of versions to fetch has overflown!".into(),
+                )
+            })?;
+
+        // Bound the number of versions to fetch by the maximum chunk size
+        num_versions_to_fetch = min(
+            num_versions_to_fetch,
+            self.max_chunk_size_for_request(config),
+        );
+
+        // Calculate the start and end versions
+        let start_version = known_version.checked_add(1).ok_or_else(|| {
+            Error::UnexpectedErrorEncountered("Start version has overflown!".into())
+        })?;
+        let end_version = known_version
```

Comment:
isn't this the same as target_version?

Reply:
This might not be same as `target_version` because we limit the number of versions to fetch in one request IIUC.

Reply:
Yep! That's the reason (see a few lines above).